### ISSUE TITLE
webpack@4 Tree Shaking doesn't eliminate code between multiple entry points in same configuration object.

### ISSUE THREAD

**Issue by ryan-roemer** [webpack/webpack#4453]
**Do you want to request a *feature* or report a *bug*?**

Bug.

**What is the current behavior?**

By way of example, I have:

- `util.js`: ES exports two methods: `red` and `blue` via two scenarios: `one-file` (all code within one file) and `re-export` (methods re-exported from external files)
- `app1.js`: Imports and uses only `red`
- `app2.js`: Imports and uses only `blue`

_Expected_: If I build webpack bundles for both apps, `app1` will have `red` and `blue` will be eliminated. Conversely, `app2` will have `blue` and `red` will be eliminated:

_Actual_: Depends on config

1. _Array of Configs_: If there is an array of configurations with a single entry point each for `app1.js` and `app2.js` the correct unused methods are dropped.
2. _Multiple Entry Points_: If there is a single configuration object with multiple entry points for both `app1.js` and `app2.js`, then **neither `blue` nor `red` are eliminated in either bundle**. I would consider this a bug. Perhaps there is some known thing for "multiple entry points get any esnext exports used by _any_ other entry point", but that seems weird.

**If the current behavior is a bug, please provide the steps to reproduce.**

We made a repository with passing and failing mocha tests to show the situation: https://github.com/FormidableLabs/webpack-tree-shaking-multiple-entry-points.

_Reproduction_:

```sh
$ git clone https://github.com/FormidableLabs/webpack-tree-shaking-multiple-entry-points.git
$ cd webpack-tree-shaking-multiple-entry-points
$ yarn install
$ yarn run build
$ yarn run test
```

The test output is:

```
  tree shaking in webpack4
    array
      one-file
        ✓ app1 should have red, not blue
        ✓ app2 should have blue, not red
      re-export
        ✓ app1 should have red, not blue
        ✓ app2 should have blue, not red
    multiple-entries
      one-file
        1) app1 should have red, not blue
        2) app2 should have blue, not red
      re-export
        3) app1 should have red, not blue
        4) app2 should have blue, not red


  4 passing (22ms)
  4 failing
```

The failing ones show that the ES methods that should be dropped aren't.

**What is the expected behavior?**

See above section on `current behavior`.

**If this is a feature request, what is motivation or use case for changing the behavior?**

Nope, it's a bug.

**Please mention other relevant information such as the browser version, Node.js version, webpack version and Operating System.**

```sh
$ node --version
v8.4.0

$  $ yarn list webpack
yarn list v1.3.2
└─ webpack@4.0.0-beta.1

$ uname -a
Darwin small.lan 16.7.0 Darwin Kernel Version 16.7.0: Thu Jan 11 22:59:40 PST 2018; root:xnu-3789.73.8~1/RELEASE_X86_64 x86_64
```

**Comment by timse** [comment]
heya, just looked into it and tried to find a solution, sadly the problems seems to be a bit bigger, as it does not only merge exports but it also shows `app1` as the issuer for both exports, therefore its not trivial to set which export is issues by whom.

need to dig deeper but not sure how deep the implications are.

**Comment by timse** [comment]
the `issuer` is a property of the module, and therefore overwritten (last one overwriting wins...)
therefore it is currently not possible to check who depends on which export, at least not in my understanding, maybe @sokra knows better?

**Comment by AlanWuji** [comment]
Any solution or advance on it now ?

**Comment by ryan-roemer** [comment]
Unfortunately not.

@TheLarkInn @sokra -- Any chance we could get this issue a little love and/or triage?

I'm also not sure what the implications are for doing common chunks / shared libs with the workaround of an array of configs w/ single entry points -- or if there are further complications straight up in tree shaking with common chunks / shared libs.

**Comment by ryan-roemer** [comment]
( _getting ahead of @sokra 's [auto-closing bot](https://twitter.com/wSokra/status/895583178634649601) on this one_ )

Unfortunately, after upgrading [my bug reproduction samples to `webpack@3.5.2`](https://github.com/FormidableLabs/webpack-tree-shaking-multiple-entry-points/commit/7e08d4dada37d163837b15ac8fabc451a02d803b) the bug still very much still exists. I'm guessing it's still a fundamental issue to webpack core.

The awkward thing with this issue, is that it's likely many folks out their aren't actually auditing their bundles to check that tree shaking is working as expected. (And in all fairness it's hard to do outside of trivial examples that you can manually inspect like I have here.)

/cc @TheLarkInn (would love some issue love on this one...)

**Comment by ryan-roemer** [comment]
Unfortunately, the latest `webpack@next` is still broken for this issue, and we're coming up on a year of it being open 😿  (`tl;dr` for folks just getting this comment -- if you have multiple entry points in a single config, code imported by **either** ends up in **both** bundles).

I've updated my examples in the error reproduction repository: https://github.com/FormidableLabs/webpack-tree-shaking-multiple-entry-points which now uses:

```js
    "webpack": "^4.0.0-beta.1",
    "webpack-cli": "^2.0.4"
```

@TheLarkInn -- Can we at least get a `bug` label on this one since I've been able to demonstrate it through every version of webpack since I opened it? For extra ❤️  would be great to actually have it triaged on the `webpack-4` track. Thanks!

**Comment by TheLarkInn** [comment]
Hmm. What is the need for the Array as entry? This causes an CJS-like merge of your modules and de-opts any hoist + shake, etc.

**Comment by ryan-roemer** [comment]
Sorry it's convoluted -- the `scenarios` array is purely for my convenience. 

Here's what **doesn't** work and what I'm guessing is the preferred way to declare entry points:

```js
module.exports = {
  entry: {
    app1: "./app1.js",
    app2: "./app2.js"
  }
}
```

That **fails** for DCE / tree shaking.

This is what **works**, although I don't like it:

```js
module.exports = ["app1", "app2"].map((name) => ({
  entry: {
    [name]: `./${name}.js`
  }
}));
```

**Comment by TheLarkInn** [comment]
Did you mean to show in your first example that they were wrapped in arrays?

**Comment by ryan-roemer** [comment]
Or to restate the bug in the simplest terms:

```js
module.exports = {
  entry: {
    app1: "./app1.js",
    app2: "./app2.js"
  }
}
```

If `app1` uses only ESM import `red`, and `app2` uses only ESM import `blue`, with both entry points in the above configuration, **both** entry points contain the code of **both** `red` and `blue`.

**Comment by edmorley** [comment]
I can reproduce this after cloning the testcase repo above (including after having run `yarn upgrade webpack webpack-cli` to upgrade from the webpack 4 beta to 4.1.1.)

It's possibly a duplicate of #5954?

**Comment by edmorley** [comment]
Or should I say #5954 is a duplicate of this, given it was filed after :-)

**Comment by joltmode** [comment]
Also stumbled on this.

@sokra on #5954 mentions to use multiple compilation passes, but that's absurd.

For instance, I'm using [webpack-manifest-plugin](https://github.com/danethurber/webpack-manifest-plugin/) and `optimization.runtimeChunk` and multi compilation messes it all up. 

On top of that, as mentioned by @loveky in #5954, with multiple compilations, there is also a problem with vendor extraction.  
Multi compilation entry points will result in multiple vendor bundles.  
It raises a question, what if rules for extraction are dynamic and lead to different vendor bundles (which webpack 4 with `splitChunks` actually has by default)? All of a sudden dead code elimination becomes dead code generation.

This seems to be a critical bug to me.

```
$ node --version
v10.1.0
```

```
$ npm list webpack
/Users/tomo/projects/mu-laravel-webpack4-typescript-react
└── webpack@4.12.0
```

**Comment by dziedrius** [comment]
We also see this issue (in webpack 4.17.0) and I agree, that most of the non SPA apps should suffer from this issue.

I'm wondering if it should end up at least in webpack 5.0 milestone, if the problem is in webpack core and can not be easily fixed.

**Comment by joltmode** [comment]
@sokra @TheLarkInn any updates on this? Any hints as to where to look to try to fix this? Anything?

**Comment by simbawus** [comment]
> Any solution or advance on it now ?

🖕

**Comment by grncdr** [comment]
Finally! This is the issue I've been scratching my head over for weeks. 🙃 

I can confirm the same problems @joltmode mentioned, as well as a new and interesting "webpack broke my build" behaviour that results.

### Scenario

We have a few different entrypoints/bundles (`application`, `admin`, and some other 1-off pages such as `login`) and with webpack4 I'm seeing a bunch of modules from the `application` entrypoint being included in the `login` bundle, even though it's **not** reachable in the import graph 

> All of a sudden dead code elimination becomes dead code generation.

### How this broke my build

Some tree-shaking is happening: certain modules that are not used in `login` are eliminated from the `login` bundle. However, some of those modules **are** referenced from top-level initialization code of modules that were included but _should_ be dead.

### Example

In `components/ComponentA.js`:

```js
import * as React from 'react'
import componentClassFactory from './componentClassFactory'
export default myComponentClassFactory({/* some config object */})
```

In `components/ComponentB.js`:
```jsx
import * as React from 'react'
export default (props) => <h1>Hello, {props.name}</h1>
```

In `components/index.js`:
```js
export { default as ComponentA } from './ComponentA'
export { default as ComponentB } from './ComponentB'
```

In `entrypoints/entrypoint1.js`:

```js
import * as React from 'react'
import {render} from 'react-dom'
import { ComponentA } from '../components'
render(<ComponentA/>, document.getElementById('app'))
```

In `entrypoints/entrypoint2.js`:

```js
import * as React from 'react'
import {render} from 'react-dom'
import { ComponentB } from '../components'
render(<ComponentB/>, document.getElementById('app'))
```

The problem occurs in **entrypoint2.js**: the import from `components/index.js` also imports `components/ComponentA.js`, which imports `components/componentClassFactory.js`, but since `components/componentClassFactory.js` is not actually _used_ anywhere from `entrypoint2`, it has been removed from the bundle. This throws an exception **at runtime when the bundle is loaded**.

**Comment by dumistoklus** [comment]
@sokra Maybe you consider to fix it in webpack5? It's a really big problem for many people. Fixing this bug can make internet faster (a little bit)

**Comment by aaarichter** [comment]
The issues also seems to occur when chunk-splitting. Even if there is only one entry in the webpack config.

```js
// chunk
export function a() { ... };
export function b() { ... };
```

```js
// app.js
...
if (condition) {
  import('a').then(module => module.a());
}
```

The generated chunk file contains both `a` and `b`

**Comment by plesiecki** [comment]
Definitely docs material.

**Comment by rmacklin** [comment]
> It's a really big problem for many people. Fixing this bug can make internet faster

To add a data point for this:

We have 258 entry points in one of our applications and compiling them using one-config-object-per-entry (rather than one config with the `entry` field resolving to all entry points) reduces our total asset size by 413KB (422897 bytes).

**Comment by targumon** [comment]
Happy issue 2nd year anniversary!
Any update?

**Comment by marcofugaro** [comment]
At the moment webpack doesn't even tree-shake unused function/variables that are declared in the same file (the `one-file` way). Not even when using `sideEffects: false` for a library.

A real-world example is the library three.js, which [exports using es6 modules but from the same file](https://github.com/marcofugaro/three.js/blob/dev/build/three.module.js).

When you import a single item `import { Vector3 } from 'three'` the whole file gets included in the bundle.

Instead importing from a file with other exports (the `re-export` way) correctly tree-shakes the library.

[Here is a test-repo that demonstrates this](https://github.com/webpack/webpack/files/3102997/three-test.zip)

**Comment by ryan-roemer** [comment]
My original error reproduction repository encompassed both "one-file" tree-shaking and "multiple file re-exports". I just updated to:

```js
"webpack": "^4.30.0",
"webpack-cli": "^3.3.1"
```

and everything still fails tree-shaking for both of those as expected: https://github.com/FormidableLabs/webpack-tree-shaking-multiple-entry-points (see [`src/one-file`](https://github.com/FormidableLabs/webpack-tree-shaking-multiple-entry-points/tree/master/src/one-file) vs. [`src/re-export`](https://github.com/FormidableLabs/webpack-tree-shaking-multiple-entry-points/tree/master/src/re-export))

Would love a webpack core team member to see if there's anything that can be done about this very, very longstanding issue. Thanks!

**Comment by theKashey** [comment]
There is another variant of this issue, this time bound to the __code splitting__.
- `A.js` imports `{X} from './B.js'`
- `A.js` dynamically `import('./C.js')`
- `C.js` imports `{Y} from './B.js'`

Results:
- `A.js` and `B.js` are in the main bundle
- `C.js` code split and forms another chunk.
- `B.js` is defined only once, and thus defined in a _maximal_ form - with X, and Y.

In this case, the problem comes from a module system - every module could be defined only once, making code splitting useless for big applications, which eventually would use all exported symbols.
And, I'm afraid, there is __no right solution__ for this - the only way to solve the problem is sub-split `B.js` to have `X-B.js` and `Y-B.js`, but as long as they might use _shared_ internals (local variables) of `B.js` this is unsafe operation, and could not be performed.

**Comment by HDC-326** [comment]
The same problem also happened to me.

I studied the source code of webpack deeply, trying to find a way out. Finally, I found a solution. I forked the test repo of @ryan-roemer and did some practice.

https://github.com/HDC-326/webpack-tree-shaking-multiple-entry-points

It seems it’s working although not very decent. It’s like a patch on gorgeous dress or someone installing windows on his mac (I didn’t do that). Just like @joltmode said, webpack is deep. I wonder if this solution covers all the cases. At least some situations were not included, such as source-map, stats and so on. The perfect solution to solve this problem needs lots of work. Maybe webpack5 would take it over.

Wish it helps.

**Comment by klis87** [comment]
I am not sure it is related, but I also have issue with tree-shaking connected with lazy loading (code splitting).

Lets say I am using a big library, like `material-ui` . Tree shaking from what I checked does work in a way that not used components are not included in production bundle. However, with lazy loading:
- we have `import { Typography } from '@material-ui/core` in application entry point
- we have `import { Dialog } from '@material-ui/core` in a component which is lazy loaded by `React.lazy`, lets say it is async chunk named `dialog`

What should happen here? material-ui library should be splitted across `main` and async `dialog` chunks. For `main` chunk ONLY `Typography` should be included, and for lazy loaded `dialog` only `Dialog` 

But what happens really? Both `Typography` and `Dialog` are included in `main-vendor`, which is a huge waste. I need `Dialog` to be in `dialog-vendor` chunk, loaded on demand. Code splitting works the best for external libraries really and if I use a very big library across multiple async chunks, I dont wanna have the whole library loaded at once, but only the parts are really needed on initial load.

I am not sure this is a feature request, or tree-shaking or code-splitting bug, but this is a serious Webpack issue in code-splitting world, causing initial apps bundles to be much bigger than they could be.

**Comment by theKashey** [comment]
@klis87 - #7782 is about is. AFAIK(as @trysound said) only rollup could handle it today.

**Comment by klis87** [comment]
@theKashey pity, Rollup doesn't have all Webpack features I guess, especially advanced HMR, so I cannot switch

Anyway, without it code splitting is rather useless because usually vendor files are much bigger than app files and really splitting vendor modules would gain the biggest benefits. Wondering whether this feature is on Webpack roadmap?

**Comment by TrySound** [comment]
@klis87 You don't need HMR for production build. Rollup is pretty good for it.

**Comment by klis87** [comment]
@TrySound but then you would need to support webpack and roolup builds within apps, which would be very unconvenient

plus, it is not only HMR which rollup is missing, there are many webpack goodies like different kinds of imports, like resolve weak etc which allows server side rendering together with code splitting which would not be possible otherwise. for instance - https://www.smooth-code.com/open-source/loadable-components/docs/server-side-rendering/ - which requires webpack plugin to work

I am sure there are much more other webpack features we don't even consider, so again, using rollup for apps is not an option for me and for many people

**Comment by sjorsvanheuveln** [comment]
I'm having the same issue. Would be great if it would be fixed.

**Comment by sjorsvanheuveln** [comment]
> Sorry it's convoluted -- the `scenarios` array is purely for my convenience to make sure that the exports coming from the same file vs. other imported files doesn't make any difference. (The scenarios confirm there is indeed no difference -- we get DCE failures both ways).
> 
> ## Single Config Fails DCE
> Here's what **doesn't** work and what I'm guessing is the preferred way to declare entry points:
> 
> ```js
> module.exports = {
>   entry: {
>     app1: "./app1.js",
>     app2: "./app2.js"
>   }
> }
> ```
> 
> That **fails** for DCE / tree shaking.
> 
> ## Array Config Works for DCE
> This is what **works**, although I don't like it:
> 
> ```js
> module.exports = ["app1", "app2"].map((name) => ({
>   entry: {
>     [name]: `./${name}.js`
>   }
> }));
> ```
> 
> And if indeed _this_ version de-optimizes and everything then we probably have an even **bigger** bug as you now have to choose between (1) DCE not working completely and (2) other de-optimizations 😉

This solved it. But I agree with @ryan-roemer. This is a big bug in webpack. I'm outputting multiple js bundles and fortunately noticed my bundles were inflating like crazy as DCE wasn't working. Thanks Ryan and hopefully this will be fixed in webpack 5! That would be great.

**Comment by Paitum** [comment]
I'm starting to architect a new application that will require many (100+) customized entry-points which will share many libraries. Like the original poster, I have verified that building them separately achieves the correct results, but doesn't scale efficiently and adds complexity.

In theory, Webpack *could* build a multi-rooted dependency-tree and then process that tree to output perfectly tree-shaken bundles per "root" or entry-point. If Webpack's dev server is running and watching all the files, then a change to a file could efficiently update the tree and update the output bundles quickly-- without requiring separate independent Webpack executions.

I'm still in early stages, but this is the wall I seem to be hitting.

**Comment by targumon** [comment]
> ```js
> module.exports = ["app1", "app2"].map((name) => ({
>   entry: {
>     [name]: `./${name}.js`
>   }
> }));
> ```

Another problem with this is that it really complicates using `webpack-merge`.
(we have slightly different webpack configurations for dev & prod, both "merge" with our 'common' configuration file.)

**Comment by hmnd** [comment]
Anyone found a better solution perhaps? webpack eats up all my ram when trying to do all my entry points individually like this.

**Comment by danoc** [comment]
I tried @ryan-roemer's repro with `webpack@5.0.0-beta.15` and the test cases still fail.

Seems like there has been an effort to improve tree shaking though:
https://github.com/webpack/changelog-v5/blob/master/README.md#nested-tree-shaking
https://github.com/webpack/changelog-v5/blob/master/README.md#inner-module-tree-shaking

**Comment by mikestreety** [comment]
Hey, I spent a few hours trying to work out a problem yesterday, only to realise it was this issue I was facing. 

If it helps:

This morning I have upgraded to `5.0.0-beta.16` (as is included in [Laravel Mix v6.0.0-alpha.1](https://github.com/JeffreyWay/laravel-mix/tree/v6.0.0-alpha.1)) and it **solved** the issue _for me_.

I haven't tested the original repo from @ryan-roemer I'm afraid.

**Comment by ryan-roemer** [comment]
Thanks for the update and glad it solved the issue for your @mikestreety !

I'm working with beta in another project, so took a moment to update the issue repository here: https://github.com/FormidableLabs/webpack-tree-shaking-multiple-entry-points/tree/chore/webpack5 (currently `5.0.0-beta.22`)

Unfortunately, it's still exhibiting the same behavior and not removing the unused code in respective bundles 😢

**Comment by ScriptedAlchemy** [comment]
Please consider updating to beta 22, the API changed and would be good to confirm this. B16 + touched a couple of areas

**Comment by vankop** [comment]
Should work with `webpack@next`. Feel free to report new issue with reproducible repo.

**Comment by ryan-roemer** [comment]
@vankop -- Yay! 🎉 

I upgraded from beta 22 to 24 in https://github.com/FormidableLabs/webpack-tree-shaking-multiple-entry-points/tree/chore/webpack5

```diff
diff --git a/package.json b/package.json
index 4c06edc..3f6ffa3 100644
--- a/package.json
+++ b/package.json
@@ -9,7 +9,7 @@
     "mocha": "^8.0.1",
     "pify": "^5.0.0",
     "uglifyjs-webpack-plugin": "^2.2.0",
-    "webpack": "^5.0.0-beta.22",
+    "webpack": "^5.0.0-beta.24",
     "webpack-cli": "^3.3.1"
   },
   "scripts": {
```

and **finally** had the comparison tests pass (for array of configs which always worked vs single configs which never did until now):

```
  tree shaking in webpack4
    array
      one-file
        ✓ app1 should have red, not blue
        ✓ app2 should have blue, not red
      re-export
        ✓ app1 should have red, not blue
        ✓ app2 should have blue, not red
    multiple-entries
      one-file
        ✓ app1 should have red, not blue
        ✓ app2 should have blue, not red
      re-export
        ✓ app1 should have red, not blue
        ✓ app2 should have blue, not red
```

Thanks!

**Comment by danoc** [comment]
So happy to hear this. Thanks!

**Comment by chiyu-git** [comment]
it's happy to hear that.
but in my project it is difficult to upgrade to the latest webpack version.
I want to know is there any negative effect if using the array of configs ?

**Comment by sokra** [comment]
You can't share chunks when using multiple configurations.

**Comment by razzp** [comment]
This bug still very much exists as of Webpack 5.36.1 when setting `optimization.runtimeChunk` to `'single'`.

**Comment by alexander-akait** [comment]
When you write `Bug`, please provide more information, otherwise it is impossible to help

**Comment by razzp** [comment]
Apologies. I've created a demo repo here: https://github.com/razzp/webpack-tree-shaking

Go get it working simply run `npm ci` followed by `npm run build` to run webpack.

Here's an outline of what's happening:

```javascript
// globals.js

const foo = 'foo';
const bar = 'bar';

export { foo, bar };
```

```javascript
// a.js

import { foo } from './globals';

console.log(foo);
```

```javascript
// b.js

import { bar } from './globals';

console.log(bar);
```

```javascript
// webpack.prod.js

module.exports = {
    entry: {
        a: './src/a.js',
        b: './src/b.js',
    }
    ...
    optimization: {
        runtimeChunk: 'single'
    }
};
```

Taking the compiled `a.js` as an example, among the boilerplate code you'll see all exports from `global.js` being included:

```javascript
... const t="foo",a="bar"
```

If you exclude the `optimization` block from the config file then `a.js` will instead compile to the following:

```javascript
(()=>{"use strict";console.log("foo")})();
```

I appreciate that code splitting will result in some extra boilterplate, however tree-shaking doesn't seem to be working at all.

**Comment by alexander-akait** [comment]
Yep, because this optimization only for production, you should not worry about it, what is the problem with this? Maybe we need update docs about it?

**Comment by razzp** [comment]
Sorry I don't follow.

Both examples (with and without `optimization.runtimeChunk` declared) are run in production mode. Surely, therefore, tree-shaking should be working in both. Am I missing something?

**Comment by alexander-akait** [comment]
Yep, bug
